### PR TITLE
[baremetal] Update `baremetal` branch with updates from `mbedtls-2.16` branch

### PR DESCRIPTION
**Pull request by sbutcher-arm**
## Description
This PR updates baremetal branch with all updates made to the `mbedtls-2.16` branch (specifically up to a59988e19749).

`baremetal` is a fork of `mbedtls-2.16`, so this PR just updates the fork with all the fixes that have since been merged into it.

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Additional comments
n/a

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

### CODE REVIEW

**Comment by AndrzejKurek** [comment]
Merge looks good, but the CI is a problem.

**Comment by sbutcher-arm** [comment]
Thanks @AndrzejKurek (and @dgreen-arm)! Am looking at the CI issue.

**Comment by sbutcher-arm** [comment]
The issue causing the CI failure will be fixed once PR #3093 has been merged, so that's a dependency.

**Comment by sbutcher-arm** [comment]
I have rebased the PR to include the merge of PR #3093. That should fix the CI problems with `zlib`.

@dgreen-arm and @AndrzejKurek - could you please re-review?

**Comment by sbutcher-arm** [comment]
PR #3093 has now been merged, so removing the 'needs preceding PR' label.

File: include/mbedtls/hmac_drbg.h [hunk 1]
```diff
@@ -146,127 +184,166 @@ int mbedtls_hmac_drbg_seed( mbedtls_hmac_drbg_context *ctx,
...
+ * \param ctx           The HMAC_DRBG context.
+ * \param additional    The data to update the state with.
+ *                      If this is \c NULL, there is no additional data.
+ * \param add_len       Length of \p additional in bytes.
+ *                      Unused if \p additional is \c NULL.
  *
  * \return              \c 0 on success, or an error from the underlying
  *                      hash calculation or
  *                      MBEDTLS_ERR_PLATFORM_FAULT_DETECTED.
- *
```
Comment (AndrzejKurek) [inline include/mbedtls/hmac_drbg.h]: This note is missing - is this intentional?

Comment (sbutcher-arm) [reply include/mbedtls/hmac_drbg.h]: This is a product of the merge, and was removed in PR #2860 to the `mbedtls-2.16` branch.

So, yes it's intentional.

**Comment by sbutcher-arm** [comment]
@AndrzejKurek - Apologies for the long delay. Could you please re-review and approve if you can? (Or you are also welcome to point out issues you missed first time round too!)

**Comment by sbutcher-arm** [comment]
Two passing reviews, and passing CI (apart from the DCO checks) so ready for merge!

**Pull request merged**